docs/developer/developing-plugins.md

```diff
@@ -0,0 +1,148 @@
+# Plugins
+
+1. Transport
+2. Handler
+3. Application
+
+## Philosophy
+Plugins should be objects with a constructor such that more than one can be 
+created if there exists more than one configuration for that plugin.
+
+# Development
+Sg-core begins by loading plugin shared object files (those that have been configured) and calling the New() function. The New() function is different for each type:
+
+Plugin Type | Initializer Function
+-|-
+Transport | `func New(* logging.Logger) transport.Transport`
+Handler | `func New() handler.MetricHandler` or `func New() handler.EventHandler`
+Application | `func New(* logging.Logger) application.Application`
+
+Both transport and application plugins contain a Run() function which encompass their primary process. Because these processes are run in a separate goroutine, a context and waitgroup are provided to synchronize with the rest of sg-core.
+
+---
+**NOTE**
+Implementations of the Run() function ___must___ begin by deferring the waitgroup Done() method, otherwise the sg-core will indefinitely hang when attempting exit.
+
+```go
+func (t *TCP) Run(ctx context.Context, wg *sync.WaitGroup, w transport.WriteFn, done chan bool) transport.Transport {
+    defer wg.Done()
```

Comment:
I will probably add this functionality to a decorator function so that plugin developers do not need to worry about it.